keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1561 @@
+- `fallback`: can handle evacuation if all previous evacuators have failed. This means that
+  `knowledgeable` or `controller` evacuators have either reported no progress, or are simply not
+  implemented for the target pod.
+
+The evacuator should first mark the pods it is interested in evacuating (either partially or fully)
+with the `evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}`
+annotation. This annotation is parsed into the `Evacuator` type in the [Evacuation API](#evacuation-api).
+
+- `EVACUATOR_CLASS`: should be unique evacuator's DNS subdomain, the maximum length is 54
+  characters (`63 - len("priority_")`)
+- `PRIORITY` and `ROLE`
+  - `controller` should always set a `PRIORITY=10000` and `ROLE=controller`
+  - other evacuators should set `PRIORITY` according to their own needs (minimum value is 0,
+    maximum value is 100000). They can use the `controller` evacuator as a reference point, if
+    they want to be run before or after the `controller` evacuator. They can also observe pod
+    annotations and detect what other evacuators have been registered for the evacuation process.
+    `ROLE` is optional and can be used as a signal to other evacuators. The `controller` value is
+    reserved for pod controllers, but otherwise there is no guidance on how the third party
+    evacuators should name their role.
+
+Example pod with evacuator annotations:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    evacuation.coordination.k8s.io/priority_fallback-evacuator.rescue-company.com: "2000"
+    evacuation.coordination.k8s.io/priority_deployment.apps.k8s.io: "10000/controller"
+    evacuation.coordination.k8s.io/priority_sensitive-workload-operator.fruit-company.com: "11000/knowledgeable-app-specific"
+  labels:
+    app: nginx
+  name: sensitive-app
+  namespace: blueberry
+```
+
+The evacuator should observe the evacuation objects that match the pods that the evacuator manages.
+It should start the evacuation only if it observes `.status.activeEvacuatorClass` that matches the
+`EVACUATOR_CLASS` it previously set in the annotation.
+
+If the evacuator is not interested in evacuating the pod, it should set
+`.status.activeEvacuatorCompleted=true`. If the evacuator is unable to respond to the evacuation,
+the `.spec.progressDeadlineSeconds` will timeout and control of the evacuation process will
+be passed to the next evacuator with the highest priority. If there is none, the pod will get
+evicted by the evacuation controller.
+
+If the evacuator is interested in evacuating the pod it should look at
+`.spec.progressDeadlineSeconds` and make sure to update the status periodically in less than that
+duration. For example, if `.spec.progressDeadlineSeconds` is set to the default value of 1800 (30m),
+it may update the status every 3 minutes. The status updates should look as follows:
+- Check that `.status.activeEvacuatorClass` still matches the `EVACUATOR_CLASS` previously set in
```

Comment:
This feels racy and brittle - easy to miss "events" ?

Reply:
`.status.activeEvacuatorClass` is only reconciled/coordinated by the the evacuation controller (single instance). 

The reaction time can be tweaked with `.spec.progressDeadlineSeconds`.

I think the following values might be reasonable, but welcome any other suggestions.

```
	// The minimum value is 600 (10m) and the maximum value is 21600 (6h).
	// The default value is 1800 (30m).
```

FYI, this mechanism is very quick if there is no evacuator advertised on the pod. It will instantly fallback to eviction.